proposals/lip-0053.md

```diff
@@ -118,43 +117,232 @@ All CCMs from terminated chains have no effect, they are neither applied nor for
+    if not all(validatorKeys[i] < validatorKeys[i + 1] for i in range(len(validatorKeys) - 1)):
+        raise Exception("Validator keys are not in lexicographic order.")
+
+    # Update validator list.
+    newActiveValidators = updateActiveValidators(validators(partnerchainID).activeValidators, ccu.params.activeValidatorsUpdate)
+    newCertificateThreshold = ccu.params.newCertificateThreshold
+    # computeValidatorsHash is defined in LIP 0058.
+    certificate = decode(certificateSchema, ccu.params.certificate)
+    if certificate.validatorsHash == bft.computeValidatorsHash(newActiveValidators, newCertificateThreshold):
+        raise Exception("Validators hash does not match.")
+```
 
-The function `uint32be(x)` returns the big endian uint32 serialization of an integer `x`, with `0 <= x < 2^32`. This serialization is always 4 bytes long.
+#### updateActiveValidators
 
-### Parameters Schema
+```python
+def updateActiveValidators(activeValidators: list[ActiveValidator], activeValidatorsUpdate: list[ActiveValidator]) -> list[ActiveValidator]:
+    validatorKeys = [validator.blsKey for validator in activeValidators]
+    for validator in activeValidatorsUpdate:
+        for currentValidator in activeValidators:
+            if currentValidator.blsKey == validator.blsKey:
+                currentValidator.bftWeight = validator.bftWeight
+        else:
+            add validator to activeValidators maintaining the entries in lexicographical order of blsKeys
+
+    #remove any entry activeValidators with bftWeight == 0
+    for validator in activeValidators:
+        if validator.bftWeight == 0:
+            activeValidators.remove(validator)
+    return activeValidators
+```
 
-Both commands will use the following `params` schema:
+#### verifyChannel
+
+```python
+def verifyChannel(ccu: CCU) -> None:
+    partnerchainID = ccu.params.sendingChainID
+    inboxTree = Merkle tree of channel(partnerchainID).inbox
+    # Update inbox by appending all cross-chain messages.
+    # crossChainMessages can be empty.
+    for ccm in ccu.params.inboxUpdate.crossChainMessages:
+        inboxTree.append(sha256(appendData))
+    newInboxRoot = inboxTree.root
+
+    # If the message witness is non-empty, update the inbox root.
+    messageWitnessHashes = ccu.params.inboxUpdate.messageWitnessHashes
+    if len(messageWitnessHashes) > 0:
+        # calculateRootFromRightWitness is specified in LIP 0031.
+        newInboxRoot = calculateRootFromRightWitness(inboxTree.size, inboxTree.appendPath, messageWitnessHashes)
+
+    # Verification for non-empty certificates.
+    if len(ccu.params.certificate) > 0:
+        outboxRootWitness = ccu.params.inboxUpdate.outboxRootWitness
+        outboxKey = MODULE_NAME_INTEROPERABILITY + SUBSTORE_PREFIX_OUTBOX_ROOT + sha256(partnerchainID)
+        proof = {
+            siblingHashes: outboxRootWitness.siblingHashes,
+            queries: [{
+                key: outboxKey,
+                value: newInboxRoot,
```

Comment:
I think the value would need to be serialized using `outboxRootSchema` defined in LIP 45.

Reply:
fixed